docs/technical/application-structure.md

```diff
@@ -0,0 +1,131 @@
+# Application Structure
+
+This document is a work in progress but covers where things can be found and
+explains how things are organized in the GitHub Desktop source.
+
+**Note:** this is an aspirational document and to be used as a starting point
+for discussions, and doesn't currently reflect reality.
+
+## `app/src`
+
+I'm going to walk through these in the order in what might look like a strange
+order, but hopefully it makes sense by the end.
+
+## Shared Modules
+
+These folders should contain modules that can be shared across any of the
+Webpack bundles that Desktop generates.
+
+ - `app/src/models` - contains the shapes used in the codebase to represent
```

Comment:
We have 3 types of models in the app:
* API
* Database
* App (the ones we use in our components and the app-store

Would be nice to have those broken out to better signal their intent.

Reply:
I don't want to get too detailed in this document about layout (especially in the `renderer` directory), but here's some thoughts:

`app/src/lib/api.ts` defines all of the API models used. I haven't explored this deeply recently but we have this sort of coupling currently to deal with:

   - `app/src/lib` -> `app/src/model` (I want this to go away)
   - `app/src/lib` -> `app/src/ui` (this is fine)

I suspect our API usage is being done in the renderer, so maybe this would live in `app/src/renderer` eventually, and then we can figure out those details. The `api.ts` module rocks in at almost 900 lines, so this is likely a great candidate to be reorganized once it finds it's home.

`app/src/lib/databases` defines these models defined inline in their respective modules. Same question about whether these should eventually live under `app/src/renderer` and can be decomposed later.

As for application models themselves, whether they live in `app/src/models` or `app/src/renderer/models` is something that can be figured out based on usage.

How much detail should we go into right now about this? We should be evolving the document as we update the layout of things anyway.

Reply:
@iAmWillShepherd bump, this looks it got swept up with me renaming the file

Reply:
I think the current level of detail is fine and I'm in favor of updating this as we figure things out. 👍